Fix #2946 : iOS Sync Integration

Brandon-T opened the pull request:
<!-- *Thank you for submitting a pull request, your contributions are greatly appreciated!* -->

## Summary of Changes
- Adds Importing/Exporting of Bookmarks to .html file so that bookmarks can be imported from Desktop or exported to Desktop.
- Adds Syncing so that Bookmarks & Folders can be synced with Desktop.
- Migrates CoreData bookmarks to Desktop/BraveCore bookmarks.
- If Migration fails, a WebPage/tab opens up with the Bookmarks.html being displayed with minor styling.
- Add BraveCore interface to iOS to choose which SyncServiceURL to use as the sync endpoint.
- Integrate BraveCore bookmarks into iOS UI.

<!-- Enter a ticket number for this PR, create a new one if it is not there yet. -->
This pull request fixes #2946, #2996, #3006, #3005, #2998, #3004, #2997, #3011, #3016 

## Submitter Checklist:

- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`

## Test Plan:
<!-- Any useful notes explaining how best to test and verify. -->


## Screenshots:
<!-- If your patch includes user interface changes that you would like to suggest or that you would like UX to look at, please include them here. -->


## Reviewer Checklist:

- [x] Issues include necessary QA labels:
  - `QA/(Yes|No)`
  - `release-notes/(include|exclude)`
  - `bug` / `enhancement`
- [x] Necessary [security reviews](https://github.com/brave/security/issues/new/choose) have taken place.
- [x] Adequate unit test coverage exists to prevent regressions.
- [x] Adequate test plan exists for QA to validate (if applicable).
- [x] Issue is assigned to a milestone (should happen at merge time).

Brandon-T committed:
- Added changes for Sync-V2 compiling.
- Joining SyncChain.
- Leaving SyncChain.
- Syncing Bookmarks.
- Migrating Bookmarks.
Fixing compiling brave core.
Adding Bookmark Interface
Got most of the UI working with some minor migration logic and wrapping of all the old logic into a new class.

iccub committed:
Move sync settings under 'general' section, add icon for it.
Adding Sync Interface Delegates that map to NSFetchedResultsViewController Delegate interface.
Fixed building for simulator and device simultaneously or separately with flags.
Fixed FavIcon not showing properly.
Moved MaterialComponents.framework as a separate dependency.
Add UI for importing and exporting bookmarks.
Add UI for migration check when joining a sync group.
Disable edit actions on permanent folders.

iccub committed:
Fix Sync screen memory leak.

Brandon-T committed:
Integrate deleting a single device from the chain by guid.
Added test code for import/exporting bookmarks from Chromium.
Added CoreData Import & Export for Bookmarks.
Updated bookmarks to use Chromium Core only and to stop using CoreData.
A migration happens at the very beginning PROGRESSIVELY and at the end displays a list of failed bookmarks (if any).
Added re-ordering of bookmarks.
Fully integrated Migration and Import/Export with the correct folder names from the iOS side, and correct nodes from the Chromium Side.
Added bookmarks interstitial page and updated API for BraveCore.
Added dynamic Bookmarks.html generation so it doesn't overwrite existing files.

Brandon-T committed:
Moved migration to the main screen instead of launch screen.

iccub reviewed (changes requested):
first look changes

Brandon-T committed:
Changes to sync text.

Brandon-T committed:
Addressing feedback.

Brandon-T committed:
Fixed nesting of bookmarks :)

Brandon-T committed:
More fixes for Bookmark Nested folders.

Brandon-T committed:
Removed unused code as we don't use this for import/export anymore.

Brandon-T committed:
Added HTTP content-security-policy and fixed rendering bugs.

iccub reviewed (changes requested):
second round, please address my remaining previous comments too

iccub reviewed (changes requested):
third round

iccub committed:
Add remaining sync strings, organize them in one place.

Brandon-T committed:
Addressed feedback.
Added migration count in case we want to limit the amount of migrations so we don't harass the user.
Added state observer in case we want to observe the bookmarks migration state from other controllers to show loading and so on.

Brandon-T committed:
Addressed feedback.

iccub reviewed (changes requested):
2 more comments, rest looks good to me

Brandon-T committed:
Fixing observers being removed before Sync Chain shuts down.
Added clean shut down of BraveCore API when application terminates so that no observer is listening during termination otherwise shutdown could be blocked.

Brandon-T committed:
Address feedback.

Brandon-T committed:
Fixed crash in moving bookmarks. Fixed favicon fetching but not from cache or from Brave-Core yet.

Brandon-T committed:
Fixed migration flag default value.

Brandon-T committed:
Changed fav_icon to load from BraveCore sync instead and only if it fails to load, then fallback to iOS FavIconFetcher.

Brandon-T committed:
Ordering bookmarks on migration

Brandon-T committed:
Added more logic to migration for deleting html file upon success.
Added more favicon logic as Brave-Core is caching icons and will not update the listener a second time.
Updated logic for exporting with order.

Brandon-T committed:
Fix favIcon crash

Brandon-T committed:
Fixed a bug where top level folders aren't being sorted during migration.

Brandon-T committed:
Import/Export will use file-system representation of the path instead of relative URL path or absolute URL path.
This fixes bugs where the user renames the file and it can't be parsed.

Brandon-T committed:
Fixed Sync not checking internet connection before leaving the sync chain.

iccub committed:
Remove BraveRewards framework, make it versionable through GH releases.

iccub committed:
Fix BraveRewards tests.

iccub committed:
Restore build_in_core script.

iccub committed:
Add most of translations(Norwegian left)

Brandon-T committed:
Prevent crash from user purposely trying to break syncing.

iccub committed:
Update brave-core-ios lib for previous commit.

iccub committed:
Do not embed Swift Standard libraries in BraveRewardsUI target.

iccub committed:
correct checksum, had to update 1.15.2 lib

iccub committed:
Add Norwegian translations.

iccub committed:
Update rewards lib to fix Rewards crash #3016
